src/OgAccess.php

```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
   }
 
   /**
+   * Checks access for entity operations on group content entities.
+   *
+   * This checks if the user has permission to perform the requested operation
+   * on the given group content entity according to the user's membership status
+   * in the given group. There is no formal support for access control on entity
+   * operations in core, so the mapping of permissions to operations is provided
+   * by PermissionManager::getEntityOperationPermissions().
+   *
+   * @param string $operation
+   *   The entity operation.
+   * @param \Drupal\Core\Entity\EntityInterface $group_entity
+   *   The group entity, to retrieve the permissions from.
+   * @param \Drupal\Core\Entity\EntityInterface $group_content_entity
+   *   The group content entity for which access to the entity operation is
+   *   requested.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The user for which to check access.
+   *
+   * @return \Drupal\Core\Access\AccessResult
+   *   The access result object.
+   *
+   * @see \Drupal\og\PermissionManager::getEntityOperationPermissions()
+   */
+  public function userAccessGroupContentEntityOperations($operation, EntityInterface $group_entity, EntityInterface $group_content_entity, AccountInterface $user = NULL) {
+    // Default to the current user.
+    if (!isset($user)) {
+      $user = $this->accountProxy->getAccount();
+    }
+
+    // Check if the user owns the entity which is being operated on.
+    $is_owner = $group_content_entity instanceof EntityOwnerInterface && $group_content_entity->getOwnerId() == $user->id();
+
+    // Retrieve the permissions and check if our operation is supported.
+    $group_entity_type_id = $group_entity->getEntityTypeId();
+    $group_bundle_id = $group_entity->bundle();
+    $group_content_bundle_ids = [$group_content_entity->getEntityTypeId() => [$group_content_entity->bundle()]];
+
+    $permissions = $this->permissionManager->getDefaultEntityOperationPermissions($group_entity_type_id, $group_bundle_id, $group_content_bundle_ids);
+
+    // Filter the permissions by operation and ownership.
+    $ownerships = $is_owner ? [FALSE, TRUE] : [FALSE];
+    $permissions = array_filter($permissions, function (GroupContentOperationPermission $permission) use ($operation, $ownerships) {
+      return $permission->getOperation() === $operation && in_array($permission->getOwner(), $ownerships);
+    });
+
+    // @todo This doesn't really vary by user but by the user's role inside of
+    //   the group. Shall we make a cache context for OgRole entities?
+    $cacheable_metadata = new CacheableMetadata;
+    $cacheable_metadata->addCacheableDependency($group_content_entity);
+    if ($user->id() == $this->accountProxy->id()) {
+      $cacheable_metadata->addCacheContexts(['user']);
+    }
+
+    if ($permissions) {
+      // Non-members might also have certain permissions within a group. Request
```

Comment:
> That's why I think we need to:
> 1. Explicitly deny access if user is blocked.

Working on this now. I've started by adding a test. It was expecting that this should already work, since in `OgAccess::userAccessGroupContentEntityOperation()` we are requesting only memberships that have `STATE_ACTIVE` or `STATE_PENDING`. Blocked users are excluded so I assumed this would default to denying access.

Unfortunately the test fails so there seems to be a problem somewhere in the logic. Investigating now.